entity-service-impl/src/main/java/org/hypertrace/entity/data/service/EntityDataServiceImpl.java

```diff
@@ -148,19 +165,26 @@ public void getAndUpsertEntities(Entities request, StreamObserver<Entity> respon
     try {
 
       Map<Key, Document> documentMap = new HashMap<>();
+      List<Entity> updatedEntities = new ArrayList<>();
       for (Entity entity : request.getEntityList()) {
         Entity normalizedEntity = this.entityNormalizer.normalize(tenantId, entity);
+        updatedEntities.add(normalizedEntity);
         Document doc = convertEntityToDocument(normalizedEntity);
         Key key = this.entityNormalizer.getEntityDocKey(tenantId, normalizedEntity);
         documentMap.put(key, doc);
       }
 
-      Streams.stream(entitiesCollection.bulkUpsertAndReturnOlderDocuments(documentMap))
-          .flatMap(document -> PARSER.<Entity>parseOrLog(document, Entity.newBuilder()).stream())
-          .map(Entity::toBuilder)
-          .map(builder -> builder.setTenantId(tenantId))
-          .map(Entity.Builder::build)
-          .forEach(responseObserver::onNext);
+      List<Entity> existingEntities =
+          Streams.stream(entitiesCollection.bulkUpsertAndReturnOlderDocuments(documentMap))
+              .flatMap(
+                  document -> PARSER.<Entity>parseOrLog(document, Entity.newBuilder()).stream())
+              .map(Entity::toBuilder)
+              .map(builder -> builder.setTenantId(tenantId))
+              .map(Builder::build)
+              .collect(Collectors.toList());
+
+      entityChangeEventGenerator.sendChangeNotification(existingEntities, updatedEntities);
```

Comment:
We need to be careful, especially on bulk upsert, about error handling. If this fails to upsert something for example, does it return it? Does it return if no change? etc. We don't want to accidentally send a delete notification if no delete actually happened.

Reply:
I'm just not sure what the error expectation is on bulk upsert in terms of atomicity and consistency. If it upserts 9 and then the 10th fails, does the whole call fail (and if so, does it roll the first 9 back?). Any changes there are out of scope, I just want to make sure we match the behavior because it's implementation-dependent.

One mismatch I can definitely see: parseOrLog. If I upsert 3 entities (all previously existing), and get the 3 originals back, but we fail to parse one, parseOrLog will log and drop it so as to not fail the whole call - whatever was malformed was overwritten, anyway. We'd end up with an array of 2 existing entities and 3 updated entities, which as I understand would send out a create event incorrectly. As to what we'd expect to happen there... I guess an update event and an unknown prior state?

Reply:
For bulkupsert, I am assuming all or none. Can validate that part.

parseOrLog is interesting. I can pull existing entities first and not depend on parseOrLog at all. That may be better.

Reply:
we use bulkwrite API on mongo side - https://docs.mongodb.com/manual/reference/method/db.collection.bulkWrite

This allows for partial writes. So we need to pull entities on task completion and use that. Correct?

Reply:
so it's kind of confusing. If it does do a partial write, let's go through a scenario.

1. Imagine I start with X, Y persisted.
2. I upsert X', Y', Z.
3. Y' fails to upsert for whatever reason. I'm not sure how this is reflected in the bulk upsert result, if at all (is Y omitted from the previous entities returned?)
   -  In the current impl, we don't use the upsert results, so would include an update of Y-> Y' that never happened (in addition to a correct creation of Z and update of X->X')
   - In the previous impl, we'd be wrong in one of two different ways, depending on what comes back in the results:
      - If Y is omitted from the previous entity result, than we'd be showing a Y' creation when Y already existed _and_ Y' failed to update
      - If Y is not omitted from previous entity result, we'd be showing a Y->Y' update that never happened. 

So this is tricky, and it does depend on the behavior of bulkwrite, but my suspicion is our bulk upsert API does not give us the info we need to completely solve this. If it omits the failed update from the previous state, we can't distinguish a failed update from a create. If it includes it, we just don't know if it succeeded or not.

Rather than make any changes in the response, I think I would want us to prevent any partial updates from happening (or verify that this is already the behavior) in doc store - all or nothing just has far less ambiguity. As long as we can trust that the updates all happened on success, then we can go back to what we had - we don't need the extra fetch and can use the prior state result of the bulk upsert (we just want to make sure that any deserialization errors of the prior state don't lead to erroneous creation notifications).

Reply:
> we just want to make sure that any deserialization errors of the prior state don't lead to erroneous creation notifications).

I just looked back and this is pretty widespread, it may be tricky to truly account for. Anywhere we call `doQuery`, underneath we do this log and drop behavior. I guess being consistent is better unless you want to try to tackle that